File: Creational/Lab1/C#/RealEstate.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection.Emit;
using System.Text;
using System.Threading.Tasks;

namespace C_
{
    public abstract class RealEstate
    {
        public int Id { get; set; }
        public string Address { get; set; }
        public string Description { get; set; }
        public double Area { get; set; }
        public double Price { get; set; }
        public string Status 
        { 
            get => _status != null ? _status : "Unknown";
            set
            {
                if (value == "New" || value == "Old" || value == "Middle aged")
                    _status = value;
                else
                    _status = "Unknown";
            } 
        }
        private string _status;

        public RealEstate(string address, string description, double area, double price, string status) { 
            Address = address;
            Description = description;
            Area = area;
            Price = price;
            Status = status;
        }

        public override string ToString(){
            return $"Id: {Id}, Address: {Address},\n Description: {Description},\nArea: {Area}, Price: {Price}\nStatus: {Status}";
        }

        public static void PrintAllInfoAboutEstates(IEnumerable<RealEstate>  realEstates){
            foreach (RealEstate realEstate in realEstates){
                Console.WriteLine(realEstate.ToString());
            }
            Console.WriteLine("\n");
        }

    }
}

File: Creational/Lab1/Python/real_estate.py
class RealEstate:
    def __init__(self, address, description, area, price, status):
        self.id = None
        self.address = address
        self.description = description
        self.area = area
        self.price = price
        self._status = "Unknown"
        self.status = status

    @property
    def status(self):
        return self._status if self._status else "Unknown"

    @status.setter
    def status(self, value):
        if value in ["New", "Old", "Middle aged"]:
            self._status = value
        else:
            self._status = "Unknown"

    def __str__(self):
        return f"Id: {self.id}, Address: {self.address},\n Description: {self.description},\nArea: {self.area}, Price: {self.price}\nStatus: {self.status}"
    
    @staticmethod
    def print_all_info_about_estates(real_estates):
        for real_estate in real_estates:
            print(str(real_estate))
